File: plugins/excel-to-sigma/skills/excel-to-sigma/scripts/lib/code_rep.py
"""Shape adapter for the Sigma WORKBOOK code representation
(POST /v2/workbooks/spec, GET|PUT /v2/workbooks/{id}/spec, POST /v2/workbooks/spec/verify).

Verified live 2026-08-03/04: this surface nests non-metadata fields under a top-level
`document` key and rejects the old flat body with HTTP 400 — including on /verify.
Sigma engineering confirmed 2026-08-03 that the DATA-MODEL code-rep surface is NOT
changing, so this adapter is workbook-only and always writes the nested shape.
Do NOT use it on /v2/dataModels/.../spec payloads — that API ignores `document`.

Reads stay tolerant of the legacy flat shape because flat artifacts still exist
on disk (committed workbook snapshots, fixtures).
"""

# Workbook elements are flat document collections; `overlays` and `panels` live
# beside them. `settings` (theme/navigation) and `agents` belong inside
# `document` too. Omitting any of these sweeps it onto the metadata envelope,
# where it is invalid and silently dropped on write.
DOC_KEYS = (
    "schemaVersion", "pages", "elements", "overlays", "panels",
    "kind", "layout", "settings", "agents",
)

# REMOVED from the API. The workbook theme is now settings.theme.name /
# settings.theme.overrides (published OpenAPI: createWorkbookSpec has zero
# occurrences of themeName/themeOverrides). The individual override keys are
# unchanged - only the container path moved. document() folds the legacy pair
# forward so specs and fixtures written before the move still produce a valid body.
LEGACY_THEME_KEYS = ("themeName", "themeOverrides")


def _fold_legacy_theme(doc, source):
    """themeName/themeOverrides -> settings.theme.{name,overrides}."""
    name = doc.get("themeName") or source.get("themeName")
    overrides = doc.get("themeOverrides") or source.get("themeOverrides")
    has_ov = isinstance(overrides, dict) and bool(overrides)
    if not name and not has_ov and not (set(LEGACY_THEME_KEYS) & set(doc)):
        return doc

    out = {k: v for k, v in doc.items() if k not in LEGACY_THEME_KEYS}
    settings = dict(out.get("settings") or {})
    theme = dict(settings.get("theme") or {})
    if name and not theme.get("name"):
        theme["name"] = name
    if has_ov:
        theme["overrides"] = {**(theme.get("overrides") or {}), **overrides}
    if not theme:
        return out
    settings["theme"] = theme
    out["settings"] = settings
    return out


def set_theme(doc, name=None, overrides=None):
    """Emitter helper - set the workbook theme in the CURRENT shape.

    Builders should call this instead of assigning the removed
    themeName/themeOverrides pair. Mutates and returns doc.
    """
    has_ov = isinstance(overrides, dict) and bool(overrides)
    if not name and not has_ov:
        return doc
    settings = doc.setdefault("settings", {})
    theme = settings.setdefault("theme", {})
    if name:
        theme["name"] = name
    if has_ov:
        theme["overrides"] = {**(theme.get("overrides") or {}), **overrides}
    return doc


def theme(spec):
    """Read the theme from either shape. Returns {"name":..., "overrides":{...}}."""
    t = (document(spec).get("settings") or {}).get("theme") or {}
    return {"name": t.get("name"), "overrides": t.get("overrides") or {}}


def document(response):
    """Return the workbook document from either the nested or legacy flat shape."""
    if not isinstance(response, dict):
        return {}
    inner = response.get("document")
    doc = inner if isinstance(inner, dict) else {
        k: v for k, v in response.items() if k in DOC_KEYS
    }
    return _fold_legacy_theme(doc, response)


def metadata(response):
    if not isinstance(response, dict):
        return {}
    return {
        k: v for k, v in response.items()
        if k != "document" and k not in DOC_KEYS and k not in LEGACY_THEME_KEYS
    }


def workbook_elements(spec):
    """Return flat elements, with read-only compatibility for old artifacts."""
    doc = document(spec)
    elements = doc.get("elements")
    if isinstance(elements, list):
        return [element for element in elements if isinstance(element, dict)]
    return [
        element
        for page in doc.get("pages", [])
        if isinstance(page, dict)
        for element in page.get("elements", [])
        if isinstance(element, dict)
    ]


def workbook_page_element_ids(spec):
    """Return {page_id: [element_id, ...]} derived from layout order."""
    import re

    result = {}
    layout = str(document(spec).get("layout") or "")
    for match in re.finditer(r'<Page\b[^>]*\bid="([^"]*)"[^>]*>(.*?)</Page>', layout, re.S):
        result[match.group(1)] = list(dict.fromkeys(
            re.findall(
                r'<(?:Element|Container|TabbedContainer|LayoutElement|GridContainer)\b'
                r'[^>]*\belementId="([^"]*)"',
                match.group(2),
            )
        ))
    return result


def workbook_page_by_element(spec):
    """Return {element_id: page_metadata}; layout is authoritative."""
    doc = document(spec)
    pages = [page for page in doc.get("pages", []) if isinstance(page, dict)]
    pages_by_id = {page["id"]: page for page in pages if page.get("id")}
    result = {}
    for page_id, element_ids in workbook_page_element_ids(doc).items():
        page = pages_by_id.get(page_id, {"id": page_id, "name": page_id})
        for element_id in element_ids:
            result.setdefault(element_id, page)
    return result


def workbook_elements_with_pages(spec):
    page_by_element = workbook_page_by_element(spec)
    return [
        (element, page_by_element.get(element.get("id") or element.get("elementId")))
        for element in workbook_elements(spec)
    ]


def _flatten_elements(doc):
    if not isinstance(doc, dict) or not isinstance(doc.get("pages"), list):
        return doc

    nested = []
    pages = []
    for page in doc["pages"]:
        page_copy = dict(page)
        nested.extend(page_copy.pop("elements", []) or [])
        pages.append(page_copy)

    elements = []
    seen = set()
    for element in list(doc.get("elements") or []) + nested:
        element_id = element.get("id") if isinstance(element, dict) else None
        if element_id and element_id in seen:
            continue
        if element_id:
            seen.add(element_id)
        elements.append(element)
    return {**doc, "pages": pages, "elements": elements}


def canonicalize_layout(layout_xml):
    """Map legacy layout aliases to the live-verified canonical tag names."""
    import re

    layout = str(layout_xml or "")
    layout = re.sub(r'<(/?)LayoutElement\b', r'<\1Element', layout)
    return re.sub(r'<(/?)GridContainer\b', r'<\1Container', layout)


def wrap(doc, extra=None):
    """Build a current request body with flat elements and canonical layout."""
    out = dict(extra or {})
    flattened = _flatten_elements(doc)
    if isinstance(flattened, dict) and "layout" in flattened:
        flattened = {**flattened, "layout": canonicalize_layout(flattened["layout"])}
    out["document"] = flattened
    return out

File: plugins/domo-to-sigma/skills/domo-to-sigma/scripts/lib/code_rep.mjs
// Shape adapter for the Sigma WORKBOOK code representation.
// Verified live 2026-08-03/04: nested `document` required on write (including
// /v2/workbooks/spec/verify); flat bodies 400. The DATA-MODEL code-rep surface is
// NOT changing — do not use this on /v2/dataModels/.../spec payloads.

// Workbook elements are flat document collections; `overlays` and `panels` live
// beside them. `settings` (theme/navigation) and `agents` belong inside
// `document` too. Omitting any of these sweeps it onto the metadata envelope,
// where it is invalid and silently dropped on write.
export const DOC_KEYS = [
  'schemaVersion', 'pages', 'elements', 'overlays', 'panels',
  'kind', 'layout', 'settings', 'agents',
];

// REMOVED from the API. The workbook theme is now settings.theme.name /
// settings.theme.overrides (published OpenAPI: createWorkbookSpec has zero
// occurrences of themeName/themeOverrides). The individual override keys are
// unchanged - only the container path moved. document() folds the legacy pair
// forward so specs and fixtures written before the move still produce a valid body.
export const LEGACY_THEME_KEYS = ['themeName', 'themeOverrides'];
export const LEGACY_HORIZONTAL_ALIGN = { start: 'left', middle: 'center', end: 'right' };
export const LEGACY_VERTICAL_ALIGN = { start: 'top', middle: 'center', end: 'bottom' };

const isObj = (v) => v !== null && typeof v === 'object' && !Array.isArray(v);

export function document(response) {
  if (!isObj(response)) return {};
  const doc = isObj(response.document)
    ? response.document
    : Object.fromEntries(Object.entries(response).filter(([k]) => DOC_KEYS.includes(k)));
  return foldLegacyTheme(doc, response);
}

// themeName/themeOverrides -> settings.theme.{name,overrides}. Returns the input
// untouched when no legacy key is present (the common, already-correct path).
function foldLegacyTheme(doc, source) {
  const name = doc.themeName || source.themeName;
  const overrides = doc.themeOverrides || source.themeOverrides;
  const hasOv = isObj(overrides) && Object.keys(overrides).length > 0;
  const hasLegacyKey = LEGACY_THEME_KEYS.some((k) => k in doc);
  if (!name && !hasOv && !hasLegacyKey) return doc;

  const out = Object.fromEntries(
    Object.entries(doc).filter(([k]) => !LEGACY_THEME_KEYS.includes(k)),
  );
  const settings = { ...(out.settings || {}) };
  const theme = { ...(settings.theme || {}) };
  if (name && !theme.name) theme.name = name;
  if (hasOv) theme.overrides = { ...(theme.overrides || {}), ...overrides };
  if (Object.keys(theme).length === 0) return out;
  settings.theme = theme;
  out.settings = settings;
  return out;
}

// Emitter helper - set the workbook theme in the CURRENT shape. Builders should
// call this instead of assigning the removed themeName/themeOverrides pair.
export function setTheme(doc, { name = null, overrides = null } = {}) {
  const hasOv = isObj(overrides) && Object.keys(overrides).length > 0;
  if (!name && !hasOv) return doc;
  doc.settings = doc.settings || {};
  doc.settings.theme = doc.settings.theme || {};
  if (name) doc.settings.theme.name = name;
  if (hasOv) {
    doc.settings.theme.overrides = { ...(doc.settings.theme.overrides || {}), ...overrides };
  }
  return doc;
}

// Read the theme from either shape.
export function theme(spec) {
  const t = (document(spec).settings || {}).theme || {};
  return { name: t.name ?? null, overrides: t.overrides || {} };
}

export function metadata(response) {
  if (!isObj(response)) return {};
  return Object.fromEntries(
    Object.entries(response).filter(
      ([k]) => k !== 'document' && !DOC_KEYS.includes(k) && !LEGACY_THEME_KEYS.includes(k),
    ),
  );
}

export function workbookElements(spec) {
  const doc = document(spec);
  if (Array.isArray(doc.elements)) return doc.elements.filter(isObj);
  return (Array.isArray(doc.pages) ? doc.pages : [])
    .filter(isObj)
    .flatMap((page) => (Array.isArray(page.elements) ? page.elements.filter(isObj) : []));
}

export function workbookPageElementIds(spec) {
  const result = {};
  const layout = String(document(spec).layout || '');
  const pagePattern = /<Page\b[^>]*\bid="([^"]*)"[^>]*>(.*?)<\/Page>/gs;
  for (const match of layout.matchAll(pagePattern)) {
    result[match[1]] = [...new Set(
      [...match[2].matchAll(
        /<(?:Element|Container|TabbedContainer|LayoutElement|GridContainer)\b[^>]*\belementId="([^"]*)"/g,
      )].map((element) => element[1]),
    )];
  }
  return result;
}

export function workbookPageByElement(spec) {
  const doc = document(spec);
  const pages = Array.isArray(doc.pages) ? doc.pages.filter(isObj) : [];
  const pagesById = Object.fromEntries(pages.filter((page) => page.id).map((page) => [page.id, page]));
  const result = {};
  for (const [pageId, elementIds] of Object.entries(workbookPageElementIds(doc))) {
    const page = pagesById[pageId] || { id: pageId, name: pageId };
    for (const elementId of elementIds) result[elementId] ||= page;
  }
  return result;
}

export function workbookElementsWithPages(spec) {
  const pageByElement = workbookPageByElement(spec);
  return workbookElements(spec).map((element) => [
    element,
    pageByElement[element.id || element.elementId],
  ]);
}

function flattenElements(doc) {
  if (!isObj(doc) || !Array.isArray(doc.pages)) return doc;
  const nested = [];
  const pages = doc.pages.map((page) => {
    const copy = { ...page };
    if (Array.isArray(copy.elements)) nested.push(...copy.elements);
    delete copy.elements;
    return copy;
  });
  const elements = [];
  const seen = new Set();
  for (const element of [...(Array.isArray(doc.elements) ? doc.elements : []), ...nested]) {
    const id = isObj(element) ? element.id : null;
    if (id && seen.has(id)) continue;
    if (id) seen.add(id);
    elements.push(element);
  }
  return { ...doc, pages, elements };
}

export function canonicalizeLayout(layoutXml) {
  return String(layoutXml || '')
    .replace(/<([/]?)LayoutElement\b/g, '<$1Element')
    .replace(/<([/]?)GridContainer\b/g, '<$1Container');
}

function canonicalizeElement(element) {
  if (!isObj(element)) return element;
  if (element.kind === 'text' && element.verticalAlign in LEGACY_VERTICAL_ALIGN) {
    return { ...element, verticalAlign: LEGACY_VERTICAL_ALIGN[element.verticalAlign] };
  }
  if (element.kind === 'kpi-chart' && isObj(element.layout)) {
    const layout = { ...element.layout };
    if (layout.anchor in LEGACY_HORIZONTAL_ALIGN) {
      layout.anchor = LEGACY_HORIZONTAL_ALIGN[layout.anchor];
    }
    if (layout.verticalAnchor in LEGACY_VERTICAL_ALIGN) {
      layout.verticalAnchor = LEGACY_VERTICAL_ALIGN[layout.verticalAnchor];
    }
    return { ...element, layout };
  }
  if (element.kind === 'tabbed-container' && isObj(element.tabBar)
      && element.tabBar.alignment in LEGACY_HORIZONTAL_ALIGN) {
    return {
      ...element,
      tabBar: {
        ...element.tabBar,
        alignment: LEGACY_HORIZONTAL_ALIGN[element.tabBar.alignment],
      },
    };
  }
  if (element.kind === 'divider' && element.align in LEGACY_VERTICAL_ALIGN) {
    const mapping = element.direction === 'vertical'
      ? LEGACY_HORIZONTAL_ALIGN
      : LEGACY_VERTICAL_ALIGN;
    return { ...element, align: mapping[element.align] };
  }
  return element;
}

function canonicalizeOverlay(overlay) {
  if (!isObj(overlay) || !isObj(overlay.drawer) || !('position' in overlay.drawer)) {
    return overlay;
  }
  const { position: _removed, ...drawer } = overlay.drawer;
  return { ...overlay, drawer };
}

export function wrap(doc, extra = {}) {
  const flattened = flattenElements(doc);
  let canonical = flattened;
  if (isObj(flattened)) {
    canonical = { ...flattened };
    if (Array.isArray(flattened.elements)) {
      canonical.elements = flattened.elements.map(canonicalizeElement);
    }
    if (Array.isArray(flattened.overlays)) {
      canonical.overlays = flattened.overlays.map(canonicalizeOverlay);
    }
    if ('layout' in flattened) canonical.layout = canonicalizeLayout(flattened.layout);
  }
  return { ...extra, document: canonical };
}
